src/firefly/java/edu/caltech/ipac/util/FitsReadUtil.java

```diff
@@ -0,0 +1,1643 @@
+         * stretch algorithm
+         */
+        if (rangeValues.getStretchAlgorithm()==RangeValues.STRETCH_ASINH) {
+            stretchPixelsUsingAsin( startPixel, lastPixel,startLine,lastLine, naxis1, imageHeader,
+                    float1dArray, pixeldata, rangeValues,slow,shigh);
+
+
+        }
+        else {
+            stretchPixelsUsingOtherAlgorithms(startPixel, lastPixel, startLine, lastLine,imageHeader.naxis1, hist,
+                    blank_pixel_value,float1dArray, pixeldata, rangeValues, slow, shigh);
+        }
+    }
+
+    private static void stretchPixelsUsingAsin(int startPixel,
+                                               int lastPixel,
+                                               int startLine,
+                                               int lastLine,
+                                               int naxis1,
+                                               ImageHeader imageHeader,
+                                               float[] float1dArray,
+                                               byte[] pixeldata,
+                                               RangeValues rangeValues,
+                                               double slow,
+                                               double shigh){
+
+        double beta = rangeValues.getBetaValue();
+        // Here we use flux instead of data since the original paper is using flux. But I don't think it is matter.
+        // flux = raw_dn * imageHeader.bscale + imageHeader.bzero, when bscale=1 and bzero=0, flux=raw_dn
+        double maxFlux = FitsReadUtil.getFlux(shigh, imageHeader);
+        double minFlux = FitsReadUtil.getFlux(slow, imageHeader);
+        if (Double.isNaN(minFlux) || Double.isInfinite((minFlux))){
+            double[] minMax=getMinMaxData(float1dArray);
+            minFlux = FitsReadUtil.getFlux(minMax[0],imageHeader);
+        }
+
+        if ( Double.isNaN(maxFlux) || Double.isInfinite((maxFlux)) ) {
+            double[] minMax=getMinMaxData(float1dArray);
+            minFlux = FitsReadUtil.getFlux(minMax[1], imageHeader);
+        }
+        int pixelCount = 0;
+        for (int line = startLine; line <= lastLine; line++) {
+            int start_index = line * naxis1 + startPixel;
+            int last_index = line * naxis1 + lastPixel;
+            for (int index = start_index; index <= last_index; index++) {
+                pixeldata[pixelCount] =(byte)  getASinhStretchedPixelValue(FitsReadUtil.getFlux(float1dArray[index], imageHeader), maxFlux, minFlux, beta);
+                pixelCount++;
+            }
+        }
+
+    }
+
+
+    public static double calculateWavelength(Header header, BinaryTableHDU tableHDU, double[] pixelCoords, String fitsType)
+            throws PixelValueException, FitsException, IOException, DataFormatException {
+
+        String algorithm =getAlgorithm(fitsType);
+        int N = header.getIntValue("WCSAXES", -1);
```

Comment:
While i was dealing with Spherex sample data and demo, i noticed that we have a problem when we were reading cube images vs. single 2d image. We can talk about it.
I think one of the problem is that NAXIS and WCS alternate is used also in a cube fits to defined the 3rd axis (not necessarily spectral) that link those images with 3rd component. 
Also, alternate WCSa, b c, etc is read by DS9 whereas CRTYPE3/CRVAL3/CDELT3 is not... 
We need to talk also in terms of extending the FITS reader with other WCS other than purely spectral coordinates. 
I would suggest that instead of getting the spectral axis from CRVAL3/CDELT3 to get it from WCSa nomenclature if it's not a cube... again, we should talk with Gregory/Vandana to understand exactly the kind of format we should support.

Reply:
The current  implementation was based on the reference paper by Greisen et al.  For ImageCube, I used both NAXIS and WCS.   The paper used CRVALia/CDELTia.  If other keywords  are used, the ticket should be updated so that I can make such changes based
on the rules.  Yes, we should know what format of FITS data we support.

Reply:
We just must have a design meeting about this before we move forward.